### DISCUSSION TITLE
Is it possible to input an array into a .effect file? How does gs_effect_set_val work?

### DISCUSSION THREAD

**Discussion by ac2522** [obsproject/obs-studio#10736] [Development Questions]
If I have `uniform float2 points[26];`, is there any way to initialize these values? 
points_cpp is a vec2[26] within my c++.

`gs_effect_set_val(pointsParam, points_cpp, sizeof(points_cpp));`
_error: Parameter 'points' set to invalid size 208, expected 8_

`gs_effect_set_vec2(pointsParam, points_cpp)` 
_error: gs_shader_set_val (GL): Size of shader param does not match the size of the input_

```
for (size_t i = 0; i < 26; ++i) {
    point_param = gs_effect_get_param_by_name(tf->shaderEffect, ("points[" + std::to_string(i) + "]").c_str());
    const vec2 point = {points_cpp[i].x, points_cpp[i].y};
    gs_effect_set_vec2(point_param, &point);
}
```
_error: effect_setval_inline: invalid param_

Maybe I've missed something, but I can't find a way to input an array, so i'm left to mannually input 26 different points. Which seems a little inefficient, especially if I want to iterate over them within the shader...

I've tried to look for examples of  working ".effect" shader files, that use arrays, but the best I could finrd was a commented array:
```
// Settings (Private)
//uniform float registerkernel[25];
```
https://github.com/Oncorporation/obs-shaderfilter/blob/master/data/examples/gaussian-blur.effect

Any help would be greatfully appreciated! I relaize there's a high chance I overlooked something completely obvious....

**Comment by Hikari-Tsai** [comment]
The effect parameter is the entire array; individual elements are not separate effect parameters. Therefore `gs_effect_get_param_by_name(effect, "points[0]")` returns null and causes the “invalid param” message. Retrieve `"points"` once and upload the contiguous array with `gs_effect_set_val`.

On current libobs, the intended pattern is:

```c
gs_eparam_t *points_param =
    gs_effect_get_param_by_name(effect, "points");

struct vec2 points[26];
/* fill points */

gs_effect_set_val(points_param, points, sizeof(points));
```

The C array must use OBS's `struct vec2` layout and be tightly contiguous. Do not pass a C++ vector object itself; for `std::vector<vec2>`, pass `points.data()` and `points.size() * sizeof(vec2)`.

The reported “expected 8” behavior came from the older effect-value path treating the parameter as one `float2`. The current libobs implementation allows the effect value buffer to resize before upload, while the shader parser retains the declared array count. If the same code still fails on a current OBS build, make sure the `.effect` file is actually the one being loaded and log the resolved parameter pointer; otherwise the plugin must either require a newer libobs or use 26 separately named uniforms as a compatibility workaround for that older version.
